diff --git a/fbandroid/java/com/facebook/catalyst/js/react-native-github/local-cli/server/checkNodeVersion.js b/fbandroid/java/com/facebook/catalyst/js/react-native-github/local-cli/server/checkNodeVersion.js
--- a/fbandroid/java/com/facebook/catalyst/js/react-native-github/local-cli/server/checkNodeVersion.js
+++ b/fbandroid/java/com/facebook/catalyst/js/react-native-github/local-cli/server/checkNodeVersion.js
@@ -36,5 +36,6 @@
       marginRight: 1,
       paddingBottom: 1,
     }));
+    process.exit(1);
   }
 };
diff --git a/fbobjc/Libraries/FBReactKit/js/react-native-github/local-cli/server/checkNodeVersion.js b/fbobjc/Libraries/FBReactKit/js/react-native-github/local-cli/server/checkNodeVersion.js
--- a/fbobjc/Libraries/FBReactKit/js/react-native-github/local-cli/server/checkNodeVersion.js
+++ b/fbobjc/Libraries/FBReactKit/js/react-native-github/local-cli/server/checkNodeVersion.js
@@ -36,5 +36,6 @@
       marginRight: 1,
       paddingBottom: 1,
     }));
+    process.exit(1);
   }
 };
