diff --git a/fbcode/logdevice/common/AllServerReadStreams.cpp b/fbcode/logdevice/common/AllServerReadStreams.cpp
--- a/fbcode/logdevice/common/AllServerReadStreams.cpp
+++ b/fbcode/logdevice/common/AllServerReadStreams.cpp
@@ -60,7 +60,7 @@
     // We actually inserted ...
 
     // Check that we aren't over capacity
-    if (streams_.size() > capacity_) {
+    if (streams_.size() > settings_->max_server_read_streams) {
       streams_.erase(insert_result.first);
       err = E::TEMPLIMIT;
       return std::make_pair(nullptr, false);
