From deadbeef Mon Sep 17 00:00:00 2001
From: yrahbar
Date: Wed, 31 Dec 1969 19:25:45 -0800
Subject: [PATCH] Test Commit

Summary: Stuff
---

diff --git a/fb-examples/tests/shipit/FBCommonFiltersTest.php-example b/fb-examples/tests/shipit/FBCommonFiltersTest.php-example
--- a/fb-examples/tests/shipit/FBCommonFiltersTest.php-example
+++ b/fb-examples/tests/shipit/FBCommonFiltersTest.php-example
@@ -141,6 +141,16 @@
     );
   }
 
+  public function testBuildDirectiveStripped(): void {
+    $in = (new ShipItChangeset())
+      ->withMessage("Summary: Test diff\nblah @build[project1,project2]");
+    $out = FBCommonFilters::applyAll($in);
+    $this->assertSame(
+      "Summary:\nTest diff\nblah",
+      $out->getMessage(),
+    );
+  }
+
   public function testAccept2ShipStripped(): void {
     $in = (new ShipItChangeset())
       ->withMessage('Summary: #accept2ship');
diff --git a/fb-examples/lib/shipit/FBCommonFilters.php-example b/fb-examples/lib/shipit/FBCommonFilters.php-example
--- a/fb-examples/lib/shipit/FBCommonFilters.php-example
+++ b/fb-examples/lib/shipit/FBCommonFilters.php-example
@@ -200,7 +200,7 @@
   ): ShipItChangeset {
     return $changeset->withMessage(
       preg_replace(
-        '/@override-unit-failures/',
+        '/(@override-unit-failures)|(@build\[.*\])/',
         '',
         $changeset->getMessage(),
       ),
--
1.7.9.5
