diff --git a/fbcode/hphp/test/slow/bad-this-closure-2.php.expectf b/fbcode/hphp/test/slow/bad-this-closure-2.php.expectf
new file mode 100644
index 0000000000000000000000000000000000000000..a0a2c0daeb0fcb1ce3a4b3a71a3f72fcc04bc65a
--- /dev/null
+++ b/fbcode/hphp/test/slow/bad-this-closure-2.php.expectf
@@ -0,0 +1,2 @@
+
+Fatal error: Don't use $this in a static method, use static:: instead in %s/bad-this-closure-2.php on line 5
diff --git "a/fbcode/hphp/test/slow/bad\\directory/i_am_a_very_bad\\file.php" "b/fbcode/hphp/test/slow/bad\\directory/i_am_a_very_bad\\file.php"
new file mode 100644
index 0000000000000000000000000000000000000000..04b1553f7b78a58997ad4e99d4864dcbfd7b7cfa
--- /dev/null
+++ "b/fbcode/hphp/test/slow/bad\\directory/i_am_a_very_bad\\file.php"
@@ -0,0 +1,7 @@
+<?hh
+
+
+<<__EntryPoint>>
+function main_i_am_a_very_bad_file() {
+echo "I'm a bad\\text\n";
+}
diff --git "a/fbcode/hphp/test/slow/bad\\directory/i_am_a_very_bad\\file.php.expect" "b/fbcode/hphp/test/slow/bad\\directory/i_am_a_very_bad\\file.php.expect"
new file mode 100644
index 0000000000000000000000000000000000000000..1ac250cb159c410d2c1625baa99447575d0f00fe
--- /dev/null
+++ "b/fbcode/hphp/test/slow/bad\\directory/i_am_a_very_bad\\file.php.expect"
@@ -0,0 +1 @@
+I'm a bad\text
